src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,246 @@
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "drivers/time.h"
+#include "flight/failsafe.h"
+#include "flight/imu.h"
+#include "flight/position.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 3);
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+);
+
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+static float getCurrentAltitude(altHoldState_s* altHoldState)
+{
+#ifdef USE_BARO
+    if (sensors(SENSOR_BARO) && baroIsCalibrated()) {
+        return 0.01f * baro.altitude;
+    }
+#endif
+    float rawAltitude = 0.01f * getEstimatedAltitudeCm();
+    if (altHoldState->smoothedAltitude == 0.0f) {
+        altHoldState->smoothedAltitude = rawAltitude;
+    }
+    float smoothFactor = 0.98f;
+    altHoldState->smoothedAltitude = (1.0f - smoothFactor) * rawAltitude + smoothFactor * altHoldState->smoothedAltitude;
+    return altHoldState->smoothedAltitude;
+}
+
+void altHoldReset(altHoldState_s* altHoldState)
+{
+    simplePidInit(&altHoldState->altPid, -50.0f, 50.0f,
+                  0.01f * altholdConfig()->altPidP,
+                  0.0f,
+                  0.01f * altholdConfig()->altPidI);
+
+    simplePidInit(&altHoldState->velPid, 0.0f, 1.0f,
+                  0.01f * altholdConfig()->velPidP,
+                  0.01f * altholdConfig()->velPidD,
+                  0.0f);
+    
+    altHoldState->throttle = 0.0f;
+    altHoldState->enterTime = millis();
+    altHoldState->exitTime = 0;
+    float externalVelocityEstimation = 0.01f * getEstimatedVario();
+    altHoldState->startVelocityEstimationAccel = altHoldState->velocityEstimationAccel - externalVelocityEstimation;
```

Comment:
Likewise, to avoid big jumps, the previous acceleration value should always be known.  If the code is not monitoring vertical acceleration in the background, this is not possible.  It may be better that the IMU continuously calculate vertical acceleration. This, and vertical velocity, could be then used directly by position hold, GPS Rescue, and other similar functions.